packages/grafana-runtime/src/utils/DataSourceWithBackend.ts

```diff
@@ -123,6 +124,10 @@ class DataSourceWithBackend<
    * Ideally final -- any other implementation may not work as expected
    */
   query(request: DataQueryRequest<TQuery>): Observable<DataQueryResponse> {
+    if (config.publicDashboardAccessToken) {
+      return publicDashboardQueryHandler(request);
+    }
+
```

Comment:
we could extract just the query method and have it behind an interface (with 2 implementations)

Reply:
so, you mean adding a new attribute in DataSourceWithBackend, of X type that needs to implement a query method and initialize it in the constructor, based on the case. 
It would be similar to what I did [here](https://github.com/grafana/grafana/pull/72056/files#diff-e66e8da9f7917f6cdf1356835e841715a074f285034ed29f6a5c7d1852fa97f7R119)

Reply:
@torkelo just wondering what you meant here with 
`we could extract just the query method and have it behind an interface (with 2 implementations)`
(changes cherry picked in [this PR](https://github.com/grafana/grafana/pull/73708) )

Reply:
I mean if we want to use a strategy like pattern, instead of the whole DataSourceAPI it would just be the query method

Reply:
Yes, I agree. I think it should be a new interface that only declarates the query method.
Maybe we can go for that approach as I did [here](https://github.com/grafana/grafana/pull/72056/files#diff-e66e8da9f7917f6cdf1356835e841715a074f285034ed29f6a5c7d1852fa97f7R119) but with a new interface instead of DataSourceApi